File: frontend/src/services/expenses.ts
import { apiDelete, apiGet, apiPost, apiPut } from '@/services/http'
import type { Expense, ExpenseFilters, ExpensePayload } from '@/types/expense'

function buildExpenseQuery(filters?: ExpenseFilters): string {
  const params = new URLSearchParams()

  if (filters?.start_date) {
    params.set('start_date', filters.start_date)
  }
  if (filters?.end_date) {
    params.set('end_date', filters.end_date)
  }
  if (filters?.category) {
    params.set('category', filters.category)
  }
  if (filters?.cash_register_session_id) {
    params.set('cash_register_session_id', filters.cash_register_session_id)
  }

  const query = params.toString()
  return query ? `?${query}` : ''
}

export async function listExpenses(
  filters?: ExpenseFilters
): Promise<Expense[]> {
  return apiGet<Expense[]>(`/api/v1/expenses${buildExpenseQuery(filters)}`)
}

export async function getExpense(expenseId: string): Promise<Expense> {
  return apiGet<Expense>(`/api/v1/expenses/${expenseId}`)
}

export async function createExpense(payload: ExpensePayload): Promise<Expense> {
  return apiPost<Expense, ExpensePayload>('/api/v1/expenses', payload)
}

export async function updateExpense(
  expenseId: string,
  payload: ExpensePayload
): Promise<Expense> {
  return apiPut<Expense, ExpensePayload>(
    `/api/v1/expenses/${expenseId}`,
    payload
  )
}

export async function deleteExpense(expenseId: string): Promise<void> {
  return apiDelete<void>(`/api/v1/expenses/${expenseId}`)
}

File: backend/app/api/v1/expenses.py
from datetime import date
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status

from app.api.dependencies import get_current_user, get_expense_service
from app.domain.expense import ExpenseCategory
from app.domain.user import User
from app.schemas.expense import CreateExpenseRequest, ExpenseResponse, UpdateExpenseRequest
from app.services.expense_service import ExpenseService

router = APIRouter(prefix="/expenses", tags=["expenses"])


@router.post("", response_model=ExpenseResponse, status_code=status.HTTP_201_CREATED)
def create_expense(
    payload: CreateExpenseRequest,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[ExpenseService, Depends(get_expense_service)],
) -> ExpenseResponse:
    expense = service.create_expense(
        description=payload.description,
        amount=payload.amount,
        category=payload.category,
        payment_method=payload.payment_method,
        created_by=current_user,
        cash_register_session_id=payload.cash_register_session_id,
        notes=payload.notes,
        expense_date=payload.expense_date,
    )
    return ExpenseResponse.from_domain(expense)


@router.get("", response_model=list[ExpenseResponse])
def list_expenses(
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[ExpenseService, Depends(get_expense_service)],
    start_date: Annotated[date | None, Query()] = None,
    end_date: Annotated[date | None, Query()] = None,
    category: Annotated[ExpenseCategory | None, Query()] = None,
    cash_register_session_id: Annotated[UUID | None, Query()] = None,
) -> list[ExpenseResponse]:
    expenses = service.list_expenses(
        actor=current_user,
        start_date=start_date,
        end_date=end_date,
        category=category,
        cash_register_session_id=cash_register_session_id,
    )
    return [ExpenseResponse.from_domain(expense) for expense in expenses]


@router.get("/{expense_id}", response_model=ExpenseResponse)
def get_expense(
    expense_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[ExpenseService, Depends(get_expense_service)],
) -> ExpenseResponse:
    expense = service.get_expense(expense_id, current_user)
    return ExpenseResponse.from_domain(expense)


@router.put("/{expense_id}", response_model=ExpenseResponse)
def update_expense(
    expense_id: UUID,
    payload: UpdateExpenseRequest,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[ExpenseService, Depends(get_expense_service)],
) -> ExpenseResponse:
    expense = service.update_expense(
        expense_id,
        description=payload.description,
        amount=payload.amount,
        category=payload.category,
        payment_method=payload.payment_method,
        actor=current_user,
        cash_register_session_id=payload.cash_register_session_id,
        notes=payload.notes,
        expense_date=payload.expense_date,
    )
    return ExpenseResponse.from_domain(expense)


@router.delete("/{expense_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_expense(
    expense_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[ExpenseService, Depends(get_expense_service)],
) -> Response:
    service.delete_expense(expense_id, current_user)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
